examples/pytorch/rgcn/link_predict_mp.py

```diff
@@ -0,0 +1,1147 @@
+                p_feats = embed_layer(p_blocks[0].srcdata[dgl.NID],
+                                      p_blocks[0].srcdata['ntype'],
+                                      p_blocks[0].srcdata['type_id'],
+                                      node_feats)
+                n_feats = embed_layer(n_blocks[0].srcdata[dgl.NID],
+                                      n_blocks[0].srcdata['ntype'],
+                                      n_blocks[0].srcdata['type_id'],
+                                      node_feats)
+
+                p_h, n_h = model(p_blocks, p_feats, n_blocks, n_feats, sample=args.sampler)
+
+                head, tail, eid = p_g.all_edges(form='all')
+                p_head_emb = p_h[head]
+                p_tail_emb = p_h[tail]
+                rids = p_g.edata['etype']
+                if queue is None:
+                    r_emb = model.w_relation[rids]
+                else:
+                    r_emb = model.module.w_relation[rids]
+                head, tail = n_g.all_edges(form='uv')
+                n_head_emb = n_h[head]
+                n_tail_emb = n_h[tail]
+            else:
+                bsize, g, p_u, p_v, rids = sample_data
+                in_feats = embed_layer(g.ndata['old_nid'],
+                                       g.ndata['ntype'],
+                                       g.ndata['type_id'],
+                                       node_feats)
+                mb_feats = model(g, in_feats, sample=args.sampler)
+                p_head_emb = mb_feats[p_u]
+                p_tail_emb = mb_feats[p_v]
+                nh_idx = th.randint(low=0, high=mb_feats.shape[0], size=(p_head_emb.shape[0],))
+                nt_idx = th.randint(low=0, high=mb_feats.shape[0], size=(p_tail_emb.shape[0],))
+                n_head_emb = mb_feats[nh_idx]
+                n_tail_emb = mb_feats[nt_idx]
+                if queue is None:
+                    r_emb = model.w_relation[rids]
+                else:
+                    r_emb = model.module.w_relation[rids]
+
+            pred_loss = get_loss(p_head_emb,
+                                 p_tail_emb,
+                                 n_head_emb,
+                                 n_tail_emb,
+                                 r_emb,
+                                 int(batch_size / chunk_size),
+                                 chunk_size,
+                                 chunk_size)
+
+            if args.sampler == 'neighbor':
+                if queue is None:
+                    reg_loss = regularization_loss(th.cat([p_feats, n_feats]), model.w_relation)
+                else:
+                    reg_loss = regularization_loss(th.cat([p_feats, n_feats]), model.module.w_relation)
+            else:
+                if queue is None:
+                    reg_loss = regularization_loss(mb_feats, model.w_relation)
+                else:
+                    reg_loss = regularization_loss(mb_feats, model.module.w_relation)
```

Comment:
If the "neighbor" and "path" -based training are so different (saw so many branches), I'd suggest just have a `run_neighbor` and `run_path`.